### PR TITLE
Apply heuristics to divine pURLs from CVERecords iff possible.

### LINKED ISSUE #763: CVE ingestion not detecting affected packages well enough
Specifically, [this CVE](https://github.com/CVEProject/cvelistV5/blob/main/cves/2024/26xxx/CVE-2024-26308.json) should result in a `commons-compress` purl.

`collectionUrl` and `packageName` should mechanically be convertible to `pkg://maven/org.apache.commons/commons-compress`, which can then semver version_range and package_status

### PR DESCRIPTION
**Pull request by bobmcwhirter**
Fixes #763

### CODE REVIEW

File: modules/ingestor/src/service/advisory/cve/divination.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+//! Helpers to try to divine pURLs from arbitrary bits of information.
+
+use cve::common::Product;
+use trustify_common::purl::Purl;
+
+pub fn divine_purl(product: &Product) -> Option<Purl> {
+    divine_maven(product)
+    // add more here as we determine the correct heuristics
```
Comment (jcrossley3) [inline modules/ingestor/src/service/advisory/cve/divination.rs:8]: Nice to have an obvious place-holder

File: modules/ingestor/src/service/advisory/cve/loader.rs [hunk 2]
```diff
@@ -216,4 +276,35 @@ mod test {
...
+        loader
+            .load(("file", "CVE-2024-26308.json"), cve, &digests)
+            .await?;
+
+        let purl = graph
+            .get_package(
+                &Purl::from_str("pkg://maven/org.apache.commons/commons-compress")?,
+                Transactional::None,
+            )
+            .await?;
```
Comment (jcrossley3) [inline modules/ingestor/src/service/advisory/cve/loader.rs:297]: Does this affirm that the `purls` field in the response from `/api/v1/vulnerability/CVE-2024-26308` will be non-empty?

**Review by jcrossley3 — APPROVED** [review]

**Comment by jcrossley3** [comment]
> Fixes #763

I referenced a couple other CVE's in that issue that this doesn't address? Should I create separate issues for them?

**Comment by bobmcwhirter** [comment]
Yah I reckon more issues.

File: modules/ingestor/src/service/advisory/cve/loader.rs (hunk 2 above)
Comment (bobmcwhirter) [reply modules/ingestor/src/service/advisory/cve/loader.rs:297]: I dunno TBH. But I hope so.

**Comment by bobmcwhirter** [comment]
We should be able to divine all if they are divinable

File: modules/ingestor/src/service/advisory/cve/loader.rs (hunk 2 above)
Comment (jcrossley3) [reply modules/ingestor/src/service/advisory/cve/loader.rs:297]: It doesn't:
```
{
  "normative": true,
  "identifier": "CVE-2024-26308",
  "title": "Apache Commons Compress: OutOfMemoryError unpacking broken Pack200 file",
  "description": "Allocation of Resources Without Limits or Throttling vulnerability in Apache Commons Compress.This issue affects Apache Commons Compress: from 1.21 before 1.26.\n\nUsers are recommended to upgrade to version 1.26, which fixes the issue.\n\n",
  "published": "2024-02-19T08:31:50.192Z",
  "modified": "2024-08-02T00:07:19.215Z",
  "withdrawn": null,
  "discovered": null,
  "released": null,
  "cwes": [
    "CWE-770"
  ],
  "average_severity": null,
  "average_score": null,
  "advisories": [
    {
      "uuid": "urn:uuid:2ea5cbf8-2f47-43ca-a234-1fe44a5641d1",
      "identifier": "CVE-2024-26308",
      "hashes": [
        "sha256:a7d8be0b6f068192e56b55a41dff77c8bc06639d60f6213ebd11078316028c37",
        "sha384:3369b48d232b6619ae41fe50022d8d460fe7fc9723ea7af58dba0b66aaafef936ac841a4a9f095806302934c81166395",
        "sha512:8d803e461a66b828ac9fbca0fe4c6fbc9a7413dd0d5a3a00aad53ee148b527a276a6dce79106263dbe1c98695f30bc3227aab3c753c8437f6fa8c6ff3662192b"
      ],
      "issuer": {
        "id": "733e416f-8d89-40c1-b162-609900c49742",
        "name": "apache",
        "cpe_key": null,
        "website": null
      },
      "published": "2024-02-19T08:31:50.192Z",
      "modified": "2024-08-02T00:07:19.215Z",
      "withdrawn": null,
      "title": "Apache Commons Compress: OutOfMemoryError unpacking broken Pack200 file",
      "labels": {
        "type": "cve"
      },
      "severity": null,
      "score": null,
      "cvss3_scores": [],
      "purls": {},
      "sboms": []
    }
  ]
}
```
Unless I'm doing it wrong. I just uploaded the one advisory, and called `/api/v1/vulnerability/CVE-2024-26308`

Comment (jcrossley3) [reply modules/ingestor/src/service/advisory/cve/loader.rs:297]: Nevermind. The branch I was testing didn't include your fix. Doh! The purl is there:

```
      "purls": {
        "affected": [
          {
            "base_purl": {
              "uuid": "41997fa4-ddb0-5def-aeba-94e8d7cd9ed6",
              "purl": "pkg://maven/org.apache.commons/commons-compress"
            },
            "version": "[1.21,1.26.0)",
            "context": null
          }
        ]
      },
```

Comment (bobmcwhirter) [reply modules/ingestor/src/service/advisory/cve/loader.rs:297]: Success!

**Pull request merged**